docs/src/lens.md

```diff
@@ -1,149 +1,251 @@
-source. For this to work, lenses actually work with closures. The real signature
-of `Lens` looks more like this (names changed for clarity):
+Now as I'm sure you've realised, the above is very inefficient. Given that we will be traversing our data very often, we need it to be cheap. (This wasn't a problem before, because when we don't need to build the inner type, we can just use references. It also wouldn't be a problem if our data was cheap to copy/clone, for example any of the primitive number types `u8`, ... `f64`.) Luckily, this is exactly the kind of thing that rust excels at. Let's rewrite the above example to be fast!
 
-```rust,noplaypen
-{{#include ../book_examples/src/lens_md.rs:lens}}
+```rust
+struct Container2 {
+    first_name: Rc<String>,
+    last_name: Rc<String>,
+    age: u16,
+}
+
+struct Name {
+    first: Rc<String>,
+    last: Rc<String>,
+}
+
+struct NameLens;
+
+impl Lens<Container2, Name> for NameLens {
+    // .. identical to previous example
+}
 ```
 
-Here `In` refers to the input to the `Lens` and `Out` is the output. `F` is a
-closure that can return a result, `R`.
+As you'll see, we've introduced `Rc`: the reference-counted pointer. You will see this and its multithreaded cousin `Arc` used pervasively in the examples. Now, the only time we actually have to copy memory is when `Rc::make_mut` is called in the `f` in `with_mut`. This means that in the case where nothing changes, all we will be doing is incrementing and decrementing reference counts. Moreover, we give the compiler the opportunity to inline `f` and `with`/`with_mut`, making this abstraction potentially zero-cost (disclaimer: I haven't actually studied the produced assembly to validate this claim).
 
-Now, instead of just being passed `Out` directly from the function, we pass the
-function a closure that will *itself* be passed an `Out`; if our closure returns
-a result, that will be given back to us.
+The trade-off is that we introduce more complexity into the `Name` type: to make changes to the data we have to use `Rc::make_mut` to get mutable access to the `String`. (The code in the lens will ensure that the newer copy of the `Rc`d data is saved to the outer type.) This means the writing fast druid code requires knowledge of the Rust pointer types (`Rc`/`Arc`, and also potentially `RefCell`/`Mutex`).
 
-This is unnecessary in the case of non-mutable access, but it is important for
-mutable access, because in many circumstances (such as when using an `Rc` or
-`Arc`) accessing a field mutably is expensive even if you don't do any mutation.
+We can actually do even better than this. Suppose that we are working on a vector of data rather than a string. We can import the `im` crate to get collections that use *structural sharing*, meaning that even when the vector is mutated, we only *Clone* what we need to. Because `im` is so useful, it is included in `druid` (behind the `im` feature).
```

Comment:
Although the introduction of `im` is important, I don't think it should be introduced in these docs. Maybe another example using the List widget would be nice. If anything I think there should be a sort of druid recipe book that demonstrates different types of ideas. i.e, different `Lens` implementations, how to use them with lists and so on.

Reply:
Yeah I had thought about having a cookbook-style list of examples at the end. I think examples aid learning concepts, as well as being useful when you have a particular problem to solve and use a particular similar example.

Reply:
Maybe the Lens page should become a Lens chapter? After you've introduced the `Lens`, you can continue with a plethora of different examples and demonstrate how they transform/map data. Even better if the examples are runnable.

Reply:
thanks for your comments @arthmis, it's super valuable to get feedback from someone who is less familiar with the subject matter!